Updating Skyblock Joining & Leaving Logic, Fixed Location updates, Added Logging.

wiviw5 opened the pull request:
This updated code is for a new bit of logic where Detection is based on what the active scoreboard name is, rather than the old system of updating based on reading a line from the scoreboard (which broke during `SKIBLOCK`).

**How it works:**
Joining to a Hypixel lobby gives us for example `Prototype` as a new name. Telling us that the client no longer is within Skyblock.

Joining into Skyblock, we receive two Scoreboards, one named `health` and the other `SBScoreboard`, this also goes along with transferring islands, we also receive this set of packets. 

We ignore `health` as this can lead to false positives for other gamemodes (As it is a common name for scoreboard names)

Next, we check for if it contains `SBScoreboard` and if it does, then we interact with the event, and we are officially "on Skyblock"

lastly, we check for any other Scoreboard, if we do not find `SBScoreboard` but another one, we interact with the Leave Skyblock event, and are "off Skyblock"

**Currently known edge cases**
One edge case of this code is that if we go straight to limbo (skipping any hub/lobby) The "leave" code is never activated, and we continue to think we are still on Skyblock, even if we leave and go to another server without a scoreboard or to a singleplayer world without a scoreboard.

So, this would affect a very minor subset of servers/singleplayer worlds, however, can be fixed with a simple IP check for `*.hypixel.net` (Not implemented) which would then only affect people who get booted straight to limbo. **Please let me know if anyone has any solutions to this.**

**Location Changes**
Updated the location change code to the function of `onScoreboardDisplay`, as it needs to update on every single Scoreboard update currently.

**Logging Changes**
Added logging for new Location change code, along with the join/leave Skyblock code.

In the future, this should be behind a toggle for developers for simply debugging if anything goes wrong with this code.

Commit:
PlayNetworkHandlerMixin.java

Previously discussed with another developer a better way to determine when a client was joining a new skyblock world. This triggers on every single new join of world. Making sure that mods get the most up-to-date info from the client.

Added some simple logging statements as well to further troubleshoot & test the aspects of Join & leave functions.

Lastly, I moved the Location updateing code up to the Objective Update class, as it functioning requires updating on every objective update received from the client.

SkyblockCore.java
Added some simple logging to Location change event registry for testing to make sure it works correctly.

Commit:
PlayNetworkHandlerMixin.java

Previously discussed with another developer a better way to determine when a client was joining a new skyblock world. This triggers on every single new join of world. Making sure that mods get the most up-to-date info from the client.

Added some simple logging statements as well to further troubleshoot & test the aspects of Join & leave functions.

Lastly, I moved the Location updateing code up to the Objective Update class, as it functioning requires updating on every objective update received from the client.

SkyblockCore.java
Added some simple logging to Location change event registry for testing to make sure it works correctly.

KyllianGamer commented:
Looks fine for the join and leave detection, given it works ofcourse. But for the logging, wouldn't devs just register a listener themselves to get a debug message like that? And furthermore, putting the actual message between "{}" doesn't make a whole lot of sense as those are generally used to log objects. Also when using the LOGGER object, the message is already prefixed with "(skyblockcore)" (the modid), so adding the name of the mod again also does not make that much sense.

wiviw5 commented:
> Looks fine for the join and leave detection, given it works ofcourse.

Haha, forgot to mention, I did some basic testing with Dungeons, Skyblock hub, Personal island, Main Lobby, and Prototype lobby and it worked flawlessly for the logic on the ones I tested here. However I need a bit more testing to make sure Kuudra, Rift, Straight to limbo, and all other Skyblock islands work, but I did not have time for those today. 

>  But for the logging, wouldn't devs just register a listener themselves to get a debug message like that?

Yes, but also, if its just built in, that allows people reporting issues/developers trying to find out why their mods not working with the core mod to do it without recompiling and re-logging into Hypixel.

> And furthermore, putting the actual message between "{}" doesn't make a whole lot of sense as those are generally used to log objects.

Only did this for visibility sake temporarily while writing this. In the future, I'd like to have something more professional yes.

> Also when using the LOGGER object, the message is already prefixed with "(skyblockcore)" (the modid), so adding the name of the mod again also does not make that much sense.

It in fact did not prefix my logs with that. Examples below:
`[23:44:17] [Render thread/INFO]: [SkyblockCore] {SKYBLOCK SB "Joined SB"} SBScoreboard`
`[23:44:17] [Render thread/INFO]: [SkyblockCore] {Health Scoreboard "Ignored"} health`
`[23:44:17] [Render thread/INFO]: [SkyblockCore] {Location Changed from: null to: Your Island}`
`[23:45:26] [Render thread/INFO]: [SkyblockCore] {Other Scoreboard, "Quit SB"} Prototype`

As to why it did not. I have no clue. Wasn't due to anything special that I am aware of either in the development environment/client testing I did there. (And yes, I was hoping it would actually prefix it, but since it didn't, I made this change to be able to spot it within my logs for the time being.)

axlecoffee committed:
Stuff

axlecoffee commented:
didnt mean to even push here but 

chAnGeS

### Changes

- Not working command handling for `/sbcore dev` handler thing yeah 
- Formatting for the logs (it looks nice I was planning to make dev toggle then push but HERE WE ARE!)

axlecoffee committed:
Revert "Stuff"

This reverts commit 13b933e78bd178a7f59db17e8fecaf81fce3a7a3.

axlecoffee committed:
Added better formatting for logs

### Changes

- Changed all of the formatting of joining/leaving skyblock and zone changes to be a bit more simplified
- Added a dev mode (currently there is no way to toggle this, to turn this off go to `src\main\...\SkyblockCore.java\` it is currently set to `TRUE`
- @wiviw5 and I talked about adding some sort of ingame command and config, I have been working on this and the pr should come soon enough

I have verifed that I am commiting to pr/12
I have checked and tested my code and it has launched and dev mode has worked both on `TRUE` and `FALSE`

axlecoffee commented:
Added better formatting for logs
### Changes

- Changed all of the formatting of joining/leaving skyblock and zone changes to be a bit more simplified
- Added a dev mode (currently there is no way to toggle this, to turn this off go to `src\main\...\SkyblockCore.java\` it is currently set to `TRUE`
- @wiviw5 and I talked about adding some sort of ingame command and config, I have been working on this and the pr should come soon enough

I have verifed that I am commiting to pr/12
I have checked and tested my code and it has launched and dev mode has worked both on `TRUE` and `FALSE`

Commit:
Fixed Logging in Main Class.

SkyblockCore.java
Temporary of course until we move these to a better class. But now the developer mode applies to all logging attempts and now looks like the other developer logs.

wiviw5 commented:
I fixed the last bit of formatting here, its of course still temporary. But now looks better for transitions between locations.

Commit:
Merge branch 'SkyBlock-Central:main' into main

axlecoffee commented:
> I fixed the last bit of formatting here, its of course still temporary. But now looks better for transitions between locations.

Thanks! I actually fixed it but when i was re-pr'ing it somehow got deleted

axlecoffee commented:
If you want you can start working on the command (or anyone else) since I can probably finish the config

Harry282 commented:
that's crazy

axlecoffee commented:
> that's crazy

👍 my phone vibrated for you, good job!

Harry282 commented:
> 👍 my phone vibrated for you, good job!

got an email for that, good job!

axlecoffee committed:
Added a config file, and formatting

## Additions
- Config creator and handler in JSON
- Added a dev toggle and location toggle
## Changes
- Merged formatting used in the last commit in this pr from @WiViW
- Changed a few logging things to include MODID

I have tested my code and it has launched with all toggles tested
I have changed where I am commiting
I have added information on how to add a config option in ModConfig.java

### TODO for this pr
- Add a command ingame to toggle
- at some point documentation...

axlecoffee commented:
Added a config file, and formatting
## Additions
- Config creator and handler in JSON
- Added a dev toggle and location toggle
## Changes
- Merged formatting used in the last commit in this pr from @wiviw
- Changed a few logging things to include MODID

I have tested my code and it has launched with all toggles tested
I have changed where I am commiting
I have added information on how to add a config option in ModConfig.java

### TODO for this pr
- Add a command ingame to toggle
- at some point documentation...

axlecoffee committed:
Remove of "Stashed changes" github

Commit:
Added a very simple command to reload from and test everything is working still.

No, it's not coded nicely. However, it works.

Also, moved the config loading & Command registry to the correct areas of the main initialization, otherwise we get the wrong information.

I also fixed the developer logger information from getting out of date data from an initial load object, to one that grabs the most up-to-date info from the loaded config. (it does not load from disk every time, just load from memory)

SkyblockCoreCommand.java

Completely new class. This is in basic the most simple reload config command I could make.

In the future, we will have different `/skyblockcore <subcommand> <options>`

Under this, it should have:

- reload (Reloads from the config file)
- nbt (Similar to the "/sba dev" or "/skytils dev nbt" commands, it grabs data from hovered items/entities. Useful for devs and curious people.)
- dev (Shows in depth data about all developer related aspects, such as the dev packets.)
- location (Shows in depth location data being handled and shown to mods)

wiviw5 commented:
Fixed some config file shenanigans & tested out a new command. (made sure that when checking if developer/location feedback mode was enabled was actually respected.)

(Note, tested on development environment & Real Client.)

### Additions

Added a local only command to do testing with, and to eventually give users feedback.

Currently looking at:
`/skyblockcore <subcommand>`

With the subcommands being:
- reload (For hot reloading of the file if manually edited.)
- nbt (Currently not implemented, for future nbt grabbing from items/entities.)
- dev (Not implemented, for future enable & disable of developer feedback logs.)
- location (Not implemented, for future enable & disable of developer feedback logs.)

### Changes

- Changed a few bits of config checking to make sure changes to the config were respected when reloaded.
- Made sure config loading & Command registration was done at the proper times in loading.

### TODO
- Better feedback to the player on the commands. (Optimally, I would like every single thing we send to the players chat window in translatable text.)
- Implementation of the commands

Commit:
Adjusted configFile configuration access.

Commit:
Added to contributors & Fixed Description.

Commit:
Removed Example Mixin

Ascynx commented on src/main/java/io/github/skyblockcore/SkyblockCore.java:
```diff
@@ -16,27 +16,96 @@
+
+
+    public static void loadConfig() {
+        File configFile = new File(FabricLoader.getInstance().getConfigDir() + File.separator + "SkyblockCoreConfig.json");
```
```suggestion
        File configFile = FabricLoader.getInstance().getConfigDir().resolve("SkyblockCoreConfig.json").toFile();
```

Ascynx commented on build.gradle:
```diff
@@ -24,11 +24,12 @@ dependencies {
 	modRuntimeOnly("me.djtheredstoner:DevAuth-fabric:1.1.2")
 	implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk8"
+	// Json thing Idk why this needs to be comented but ill do it like this since orginization -axle
+	implementation 'com.google.code.gson:gson:2.8.9'
```
Isn't Gson already included with fabric?

Ascynx commented on src/main/java/io/github/skyblockcore/SkyblockCore.java:
```diff
@@ -16,27 +16,96 @@
+        config.setLocation(false); // this is used for location output (see line 111 - 112) this is used for debugging new locations.
+        Gson gson = new GsonBuilder().setPrettyPrinting().create();
+        String json = gson.toJson(config);
+        File configFile = new File(FabricLoader.getInstance().getConfigDir() + File.separator + "SkyblockCoreConfig.json");
```
See above comment (in SkyblockCore)

Ascynx commented on src/main/java/io/github/skyblockcore/command/SkyblockCoreCommand.java:
```diff
@@ -0,0 +1,51 @@
+
+    private static int location() {
+        // TODO Implement config editing & changing of the variable in config.
+        LOGGER.info(TITLE + " This would of started/stopped showing developer location data.");
```
```suggestion
        LOGGER.info(TITLE + " This would have started/stopped showing developer location data.");
```

Ascynx commented on src/main/java/io/github/skyblockcore/command/SkyblockCoreCommand.java:
```diff
@@ -0,0 +1,51 @@
+
+    private static int dev() {
+        // TODO Implement config editing & changing of the variable in config.
+        LOGGER.info(TITLE + " This would of started/stopped showing developer data.");
```
```suggestion
        LOGGER.info(TITLE + " This would have started/stopped showing developer data.");
```

Ascynx commented on src/main/java/io/github/skyblockcore/command/SkyblockCoreCommand.java:
```diff
@@ -0,0 +1,51 @@
+    private static int nbt() {
+        // TODO Implement NBT Handling Logic here
+        //  This will be similar to "/sba dev" or "/skytils dev nbt".
+        LOGGER.info(TITLE + " This would of started/stopped the NBT data.");
```
```suggestion
        LOGGER.info(TITLE + " This would have started/stopped the NBT data.");
```

Ascynx commented on src/main/java/io/github/skyblockcore/SkyblockCore.java:
```diff
@@ -16,27 +16,96 @@
     public static final Logger LOGGER = LoggerFactory.getLogger(ModID);
 
+    private static ModConfig config;
+    private static final String TITLE = "[SkyblockCore]";
```
Logger already prefixes what is sent through it with "(skyblockcore)" I don't see the use.

axlecoffee replied:
its cooler to use []

Ascynx commented on src/main/java/io/github/skyblockcore/mixin/PlayNetworkHandlerMixin.java:
```diff
@@ -15,38 +14,31 @@
+    private ClientWorld world;
+    @Shadow
+    @Final
+    private static final Logger LOGGER = LoggerFactory.getLogger(ModID);
```
Just call it from SkyblockCore when used, why create a new one?

Ascynx commented on src/main/java/io/github/skyblockcore/mixin/PlayNetworkHandlerMixin.java:
```diff
@@ -64,4 +56,46 @@ void onScoreboardDisplay(ScoreboardDisplayS2CPacket packet, CallbackInfo ci) {
+
+    private static final String SKYBLOCK_SCOREBOARD = "SBScoreboard";
+    private static final String HEALTH_SCOREBOARD = "health";
+    private static final String TITLE = "[SkyblockCore]";
```
See first comment in SkyblockCore

axlecoffee committed:
Update src/main/java/io/github/skyblockcore/SkyblockCore.java

Co-authored-by: Ascynx <78341107+Ascynx@users.noreply.github.com>

Commit:
Update src/main/java/io/github/skyblockcore/command/SkyblockCoreCommand.java

Co-authored-by: Ascynx <78341107+Ascynx@users.noreply.github.com>

Commit:
Update src/main/java/io/github/skyblockcore/command/SkyblockCoreCommand.java

Co-authored-by: Ascynx <78341107+Ascynx@users.noreply.github.com>

Commit:
Update src/main/java/io/github/skyblockcore/command/SkyblockCoreCommand.java

Co-authored-by: Ascynx <78341107+Ascynx@users.noreply.github.com>

Commit:
Fixes requested by Ascynx.

wiviw5 commented:
Fixes requested by Ascynx included fixing the loggers in separate classes, some spelling, removal of GSON (provided by fabric already), and questions on logger reasoning.

Harry282 commented:
Why is the config instance nullable? There shouldn't have to be a null check every time we want to access the config.

axlecoffee commented:
Oh that's odd, I think my IDE complained without it... but yeah you're right it shouldn't be nullable

axlecoffee commented:
If you want @Harry282 you can commit a fix

TGWaffles committed:
fix and optimise sub-island location detection from scoreboard

Commit:
Merge pull request #1 from TGWaffles/fix-location

fix and optimise sub-island location detection from scoreboard

TGWaffles commented:
Fixed the (sub-island) location listener.

It used to listen only for objective updates, but the location isn't actually an objective name - instead it's a team name.

Now we only update the location when we actually receive a team update packet from Hypixel with the location character (U+23E3) as the first character (excluding whitespace). 

Works very reliably, and updates instantly when the player location changes in scoreboard. Added bonus effect of partial location updates as Hypixel partially updates the location (but these only persist for a few milliseconds until Hypixel sends the full update).

axlecoffee committed:
Added better logging, chat stuff and string public

### Changes
- Made the TITLE string public
- Made scoreboard strings public
- Removed the private versions of ^
- Removed some random lines blankness
- Formatted a tiny bit of logging and changed a few things
### Additions
- Commands now have ingame replies (No color yet, and just basic they still dont do stuff)

### TODO
- Make commands do stuff
- Add color to commands
- change around some logging
- Fix config being nullable
- add a time calc for reloading config
- add a way to reload config
- fix all of the headaches with LOGGER sometimes showing weird stuff
- Fix issues with location

I have checked that my code works in developer launch and normal minecraft

axlecoffee commented:
### Changes
- Made the TITLE string public
- Made scoreboard strings public
- Removed the private versions of ^
- Removed some random lines blankness
- Formatted a tiny bit of logging and changed a few things
### Additions
- Commands now have ingame replies (No color yet, and just basic they still dont do stuff)

### TODO
- Make commands do stuff
- Add color to commands
- change around some logging
- Fix config being nullable
- add a time calc for reloading config
- add a way to reload config
- fix all of the headaches with LOGGER sometimes showing weird stuff
- Fix issues with location

I have checked that my code works in developer launch and normal minecraft
![image](https://github.com/SkyBlock-Central/SkyBlock-Core/assets/95949634/fb891ebf-3c7d-4ef3-b51c-32125b00abb8)

Commit:
Fixing up the player checks.

wiviw5 commented:
Fixed up player checks.

TGWaffles commented on src/main/java/io/github/skyblockcore/SkyblockCore.java:
```diff
@@ -16,27 +16,97 @@
+            }
+        } else {
+            createConfig();
+            LOGGER.info(TITLE + " Config successfully created! Welcome to SkyblockCore, a Skyblock mod designed for modern Minecraft.");
```
Is SkyblockCore... a mod?

TGWaffles replied:
If this is just temporary for testing that's fine, maybe add a comment to specify that

axlecoffee replied:
What might end up happening is that itll download on start, not sure yet

TGWaffles commented on src/main/java/io/github/skyblockcore/SkyblockCore.java:
```diff
@@ -16,27 +16,97 @@
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
```
Might be an idea to move this to the ModConfig class itself so we don't have to deal with file logic in the main class.

axlecoffee replied:
I am planning on doing that

TGWaffles commented on src/main/java/io/github/skyblockcore/command/SkyblockCoreCommand.java:
```diff
@@ -0,0 +1,63 @@
+            LOGGER.info(TITLE + " NBT data has been copied to your clipboard!");
+        }
+//      If thomas wants to do smth better here he can.
+        if (MinecraftClient.getInstance().player == null) return 0;
```
```suggestion
        if (MinecraftClient.getInstance().player == null) return 0;
```

This is good.

TGWaffles commented on src/main/java/io/github/skyblockcore/mixin/PlayNetworkHandlerMixin.java:
```diff
@@ -1,67 +1,81 @@
             }
+            LOGGER.info(TITLE + " Leaving Skyblock...");
+            if (onSkyblock) LeaveSkyblockCallback.EVENT.invoker().interact();
         }
```
Can this not either be an else to the previous if statement, or a check after "health" with an early-return?

TGWaffles committed:
Update src/main/java/io/github/skyblockcore/command/SkyblockCoreCommand.java

Remove unneeded meme comment.

Commit:
Changed around format of getting whether the user has joined or left skyblock.

This removes a single string comparison now that we check for anything but specific strings associated with skyblock.

Commit:
Merge remote-tracking branch 'origin/main' into main

wiviw5 commented:
Changed around format of getting whether the user has joined or left Skyblock.

This removes a single string comparison now that we check for anything but specific strings associated with Skyblock.

Commit:
Merge branch 'SkyBlock-Central:main' into main

axlecoffee committed:
I commit first so I dont deal with merge confits

axlecoffee commented:
### Additions 
- Made a new class for config stuff to remove it from the main class
- `io.github.skyblockcore.event.ConfigManager`
- Made all the commands (aside from NBT) do stuff 

### Changes
- Changed all old config uses to use the new method 
![image](https://github.com/SkyBlock-Central/SkyBlock-Core/assets/95949634/e180e636-64b6-4229-8d72-caf3ce4a8edf)

### Known issues
- I somehow caused see image

Commit:
Added NBT Access. (Copying was not working, eventually needs to be added)

axlecoffee committed:
idk this might not build

axlecoffee commented:
### Additions 
- Log filtering 
- log filter toggle command
- log filter config lines
- test command!
- cute catboys

### Changes
- Removed the femboys from an old update as it was ruining the motivation of the team 

I have tested this and it has launched and built in dev and in minecraft 

### TODO
- Fix config not allowing new code to be printed without hard reset 
- colors 🌈 🏳️‍🌈 🏳️‍⚧️ 👑 💅 💅 💅 💅 👄 💋 💄

Commit:
Ty Ascynx for making this known. Copying now works.

Commit:
Merge remote-tracking branch 'origin/main' into main

TGWaffles commented:
Is this ready for reviews?

axlecoffee commented:
> Is this ready for reviews?

You have pending changes iirc

TGWaffles commented on src/main/java/io/github/skyblockcore/SkyblockCore.java:
```diff
@@ -47,8 +79,39 @@ public void onInitializeClient() {
             return ActionResult.PASS;
         }));
+        // Source from https://github.com/apace100/show-me-what-you-got for implementation of the mixin.
+        ClientTickEvents.START_CLIENT_TICK.register(tick -> {
+            if (!NBTCOPYING) return;
+            MinecraftClient client = MinecraftClient.getInstance();
+            if (client.player != null && client.currentScreen instanceof HandledScreen) {
+                HandledScreenFocusedSlotAccessor focusedSlotAccessor = (HandledScreenFocusedSlotAccessor) client.currentScreen;
+                Slot focusedSlot = focusedSlotAccessor.getFocusedSlot();
+                boolean isCtrlPressed = InputUtil.isKeyPressed(MinecraftClient.getInstance().getWindow().getHandle(), GLFW.GLFW_KEY_RIGHT_CONTROL);
+                if (isCtrlPressed) {
+                    if (client.player.currentScreenHandler.getCursorStack().isEmpty() && focusedSlot != null && focusedSlot.hasStack()) {
+                        ItemStack itemToCopyNBT = focusedSlot.getStack();
+                        if (itemToCopyNBT.getNbt() == null) return;
+                        String itemNBT = "minecraft:" + itemToCopyNBT.getItem().getName().getString() + " " + itemToCopyNBT.getNbt();
+                        if (ConfigManager.getConfig() != null && ConfigManager.getConfig().isDev()) {
+                            LOGGER.info(TITLE + " [Dev NBT] > " + itemNBT);
+                        }
+                        MinecraftClient.getInstance().keyboard.setClipboard(itemNBT);
+                    }
+                }
+            }
+        });
```
Putting this in client tick seems like a REALLY bad idea. Would this not be better eg in tooltip event so it's only triggered in a gui and when we're hovering over an item? 

Could possibly also just mixin over the method that colours the focused slot?

CraftedFury replied:
This will be opened as an issue.
